src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.ValueTaskSource.cs

```diff
@@ -0,0 +1,248 @@
+            internal short Version => _source.Version;
+
+            internal void RegisterForCancellation(CancellationToken cancellationToken)
+            {
+#if DEBUG
+                Debug.Assert(cancellationToken.CanBeCanceled);
+                Debug.Assert(!_cancellationHasBeenRegistered, "Cannot register for cancellation twice");
+                _cancellationHasBeenRegistered = true;
+#endif
+
+                // Quick check to make sure the IO hasn't completed
+                if (_overlapped != null)
+                {
+                    // Register the cancellation only if the IO hasn't completed
+                    long packedResult = Interlocked.CompareExchange(ref _result, TaskSourceCodes.RegisteringCancellation, TaskSourceCodes.NoResult);
+                    if (packedResult == TaskSourceCodes.NoResult)
+                    {
+                        _cancellationRegistration = cancellationToken.UnsafeRegister((s, token) => Cancel(token), this);
+
+                        // Switch the result, just in case IO completed while we were setting the registration
+                        packedResult = Interlocked.Exchange(ref _result, TaskSourceCodes.NoResult);
+                    }
+                    else if (packedResult != TaskSourceCodes.CompletedCallback)
+                    {
+                        // Failed to set the result, IO is in the process of completing
+                        // Attempt to take the packed result
+                        packedResult = Interlocked.Exchange(ref _result, TaskSourceCodes.NoResult);
+                    }
+
+                    // If we have a callback that needs to be completed
+                    if ((packedResult != TaskSourceCodes.NoResult) && (packedResult != TaskSourceCodes.CompletedCallback) && (packedResult != TaskSourceCodes.RegisteringCancellation))
+                    {
+                        CompleteCallback((ulong)packedResult);
+                    }
+                }
+            }
+
+            internal virtual void ReleaseNativeResource()
+            {
+                // Ensure that cancellation has been completed and cleaned up.
+                _cancellationRegistration.Dispose();
+
+                // Free the overlapped.
+                // NOTE: The cancellation must *NOT* be running at this point, or it may observe freed memory
+                // (this is why we disposed the registration above).
+                if (_overlapped != null)
+                {
+                    _strategy._fileHandle.ThreadPoolBinding!.FreeNativeOverlapped(_overlapped);
+                    _overlapped = null;
+                }
+
+                // Ensure we're no longer set as the current AwaitableProvider (we may not have been to begin with).
+                // Only one operation at a time is eligible to use the preallocated overlapped
+                _strategy.CompareExchangeCurrentOverlappedOwner(null, this);
```

Comment:
I would encourage you to look at reusing the ValueTaskSource instance as part of this change: that's the primary reason you'd move to this model in the first place, otherwise you might as well just stick with the task code. This code here is already managing renting/returning a reusable object; doing so with VTS shouldn't end up costing anymore as you'll just end up renting/returning that instead.

Reply:
Thank you. Makes sense to add the caching logic in this PR.